File: java/src/main/java/com/bidyut/tech/crawler/Droid1.java
package com.bidyut.tech.crawler;

import org.jetbrains.annotations.NotNull;

import java.net.URI;
import java.util.*;

public class Droid1 extends Droid {
    @Override
    public @NotNull String getName() {
        return "Droid1";
    }

    @Override
    public @NotNull List<String> invoke(@NotNull String uri, int depth) {
        debugLog("Starting to crawl...");
        final Queue<Entry> queue = new LinkedList<>();
        queue.add(new Entry(uri, depth));
        final List<String> uris = new LinkedList<>();
        final Set<URI> visited = new HashSet<>();
        while (!queue.isEmpty()) {
            final Entry entry = queue.poll();
            uris.add(entry.uri.toString());
            if (entry.depth > 0 && !visited.contains(entry.uri)) {
                visited.add(entry.uri);
                crawlAndQueue(entry, queue::add).join();
            }
        }
        shutdown();
        return uris;
    }
}

File: kotlin/src/main/kotlin/Droid1.kt
import io.ktor.http.*
import kotlinx.coroutines.runBlocking

class Droid1(
    debugLogEnabled: Boolean = false
) : Droid(debugLogEnabled) {
    override val name = "Droid1"

    override fun invoke(uri: String, depth: Int): List<String> {
        debugLog("Starting to crawl...")
        val queue = mutableListOf<Entry>()
        queue.add(Entry(Url(uri), depth))
        val uris = mutableSetOf<String>()
        val visited = mutableSetOf<String>()
        while (queue.isNotEmpty()) {
            val entry = queue.removeAt(0)
            uris.add(entry.uri.toString())
            if (entry.depth > 0 && !visited.contains(entry.uri.toString())) {
                visited.add(entry.uri.toString())
                runBlocking { crawlAndQueue(entry, queue::add) }
            }
        }
        return uris.toList()
    }
}
